### ISSUE TITLE
clang-tidy readability-else-after-return is not compatible with C++17 if init-statements

### ISSUE THREAD

**Issue by matiaslindgren** [llvm/llvm-project#93491]
The following code emits a false positive warning on trunk clang-tidy.
```c++
int f();

int g() {
    if (auto x{f()}; x > 0) {
        return x;
    } else {
        return -x;
    }
}
```
`x` does not exist outside the scope of the `if` and `else` branches so the `else` branch cannot be removed in this case. Moving the init-statement outside the `if` statement would change the logic too much because `x` lives now in the outer scope.

**Comment by dwblaikie** [comment]
> Moving the init-statement outside the if statement would change the logic too much because x lives now in the outer scope.

my 2c is that I disagree with ^ - since the else is the rest of the code in this function (necessarily, since both if and else return), there's no increase in scope by moving the variable declaration outside the `if` and that'd be OK and consistent with LLVM's style guide, I think...

**Comment by matiaslindgren** [comment]
> my 2c is that I disagree with ^ - since the else is the rest of the code in this function (necessarily, since both if and else return), there's no increase in scope by moving the variable declaration outside the `if` and that'd be OK and consistent with LLVM's style guide, I think...

Contrived counter-example: if `x` is already shadowing an `x` from the outer scope I would have to move the outer `x` to a new scope too (and possibly repeat for all parent scopes). Then again, shadowing is a readability issue by itself.

In my simple case I was able to do exactly as you suggested, but I could imagine situations where I'd prefer to turn off `readability-else-after-return` rather than refactor.

**Comment by dwblaikie** [comment]
Fair - though I'm not sure turning it off would be a big deal. It's a somewhat esoteric LLVM convention to begin with, so I wouldn't expect it to be a thing a lot of other projects use to the point of feeling strongly about it.

I don't totally object to some changes being made here, but usually in LLVM code itself, I'd probably advocate for the change that is being suggested here - moving the variable out of the `if` and dropping the `else`.

**Comment by matiaslindgren** [comment]
Makes sense, I'm ok with closing this if disabling `readability-else-after-return` is the preferred way to go.

**Comment by dwblaikie** [comment]
Yeah, i'm not probably the best decider here/now a maintainer of clang-tidy in particular, so I think we can leave it open for other folks to chime in.

**Comment by HerrCai0907** [comment]
It should already be handled. I will check more detail.

**Comment by HerrCai0907** [comment]
You can use `readability-else-after-return.WarnOnConditionVariables = false` to disable warning for the conditional variables.
